src/client_builder.rs

```diff
@@ -141,10 +266,62 @@ where
         let tls = handshake(self.domain.as_ref(), tcp)?;
 
         let mut client = Client::new(tls);
-        if !self.starttls {
+        if self.mode != ConnectionMode::StartTls {
```

Comment:
I think now that we have `self.mode`, I would expect this to _also_ respect `ConnectionMode::Tls`. That is, if someone did:
```rust
builder.tls().connect_with(|_host, tcp| tcp)
```
I would expect them to end up with a TLS connection. This comes back to the fact that `connect_with` is for wrapping the channel _in any way_, and so does not in and of itself imply that TLS has already been set up. Which also then means that it'd be a mistake to call `.tls().connect_with(|_, tcp| Rustls::new(tcp))`, because you'd end up running TLS-over-TLS.

Reply:
We should also document this behavior.

Reply:
Which that now greatly complicates this method's use.  As if you go with that logic how would StartTLS behave?  Would the user of this method have to initialize starttls? or would the function still perform that behavior?

As your original intent for this method was to "customize" the TLS setup, which one cannot do AFTER the TLS object has been created and wrapped around the raw TCP connection.  Thus this is intended to be done before wrapping.

So, perhaps we should just scrap this method entirely until we come up with a better way to customize the connections?  e.g. a customize_native / customize_rustls ?  that provides the config/builder struct allowing the user to tweak it ?

e.g.
```rust
client_builder.tls().customize_nativetls(|&mut builder| .... modify builder).connect()?;
```

and in our build_native_tls() internal method we do this.

```rust
fn build_tls_native(&self, tcp: TcpStream) -> Result<ConnType> {
        let mut builder = NativeTlsConnector::builder();
        if self.no_tls_verify {
            builder.danger_accept_invalid_certs(true);
            builder.danger_accept_invalid_hostnames(true);
        }
        if Some(customizer) = self.customize_nativetls_callback {
           customizer(builder);
        }
        let ssl_conn = builder.build()?;
        Ok(Box::new(NativeTlsConnector::connect(
            &ssl_conn,
            self.domain.as_ref(),
            tcp,
        )?))
    }
```

We can either have two separate customization hooks OR have a unified with that takes an enum of either the nativetls builder or the rusttls config.   Also if a consumer is customizing the TLS connection do we want to process the defaults always? (loading stock certs?) or add a config option to disable that functionality?

Reply:
Hmm, I do see your point, but at the same I simultaneously think this method is useful as-is _and_ I think it should do something reasonable in the `client.tls().connect_with` situation. I suppose we could return a runtime error? An alternative would be to do a type-state kind of thing where instead of keeping `mode` as a variable we keep it as a type parameter, but that feels excessive. Maybe just document that it'll return an error if `tls()` is set?

Reply:
What error should it return?  ( a new variant?)

Reply:
Yeah, a new variant seems fine